adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,190 @@
+# frozen_string_literal: true
+
+# Copyright 2020 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'opentelemetry/trace/status'
+
+require_relative '../util/queue_time'
+
+module OpenTelemetry
+  module Adapters
+    module Rack
+      module Middlewares
+        # TracerMiddleware propagates context and instruments Rack requests
+        # by way of its middleware system
+        class TracerMiddleware # rubocop:disable Metrics/ClassLength
+          class << self
+            def allowed_rack_request_headers
+              @allowed_rack_request_headers ||= Array(config[:allowed_request_headers]).each_with_object({}) do |header, memo|
+                memo["HTTP_#{header.to_s.upcase.gsub(/[-\s]/, '_')}"] = build_attribute_name('http.request.headers.', header)
+              end
+            end
+
+            def allowed_response_headers
+              @allowed_response_headers ||= Array(config[:allowed_response_headers]).each_with_object({}) do |header, memo|
+                memo[header] = build_attribute_name('http.response.headers.', header)
+                memo[header.to_s.upcase] = build_attribute_name('http.response.headers.', header)
+              end
+            end
+
+            def build_attribute_name(prefix, suffix)
+              prefix + suffix.to_s.downcase.gsub(/[-\s]/, '_')
+            end
+
+            def config
+              Rack::Adapter.instance.config
+            end
+
+            private
+
+            def clear_cached_config
+              @allowed_rack_request_headers = nil
+              @allowed_response_headers = nil
+            end
+          end
+
+          EMPTY_HASH = {}.freeze
+
+          def initialize(app)
+            @app = app
+          end
+
+          def call(env)
+            original_env = env.dup
+            extracted_context = OpenTelemetry.propagation.extract(env)
+            frontend_context = create_frontend_span(env, extracted_context)
+
+            # restore extracted context in this process:
+            OpenTelemetry::Context.with_current(frontend_context || extracted_context) do
+              request_span_name = create_request_span_name(env['REQUEST_URI'] || original_env['PATH_INFO'])
+              tracer.in_span(request_span_name,
+                             attributes: request_span_attributes(env: env),
+                             kind: :server) do |request_span|
+                @app.call(env).tap do |status, headers, response|
+                  set_attributes_after_request(request_span, status, headers, response)
+                end
+              end
+            end
+          ensure
+            finish_span(frontend_context)
+          end
```

Comment:
This seems pretty convoluted and incorrect. If I understand correctly, with the "frontend" span enabled, we'll end up with 2 spans nested like:
```yaml
name: http_server.queue
start_timestamp: T
end_timestamp: T+D
kind: server
attributes:
  - start_time: request_start_time
children:
  - name: $REQUEST_URI
    start_timestamp: T-ish
    end_timestamp: T+D-ish
    kind: server
    ...
```
There are several problems here:
1. The duration of the 2 spans is about the same, so the queue time is only represented in an attribute of the outer span. It is more useful to create the `http_server.queue` span with `start_timestamp: request_start_time`, which will more accurately represent the LB request duration as the span's duration.
1. Both spans are marked as `kind: server`, which is logically incorrect and would be reported as so by some vendors' trace quality reports. This is supposed to represent time between a proxy/LB receiving a request and the service handling it, there has been chatter about representing this with something like `kind: proxy` (https://github.com/open-telemetry/opentelemetry-specification/issues/51 and https://github.com/open-telemetry/opentelemetry-specification/pull/337).
1. For analysis backends deriving SLIs from trace data, it is possible they may count requests twice if both spans are `kind: server`.
1. Calling the frontend span `http_server.queue` is inaccurate - it actually represents (almost) the entire request processing time from the perspective of the LB/proxy.

Alternatives:
1. Rename the frontend span to `http_server.proxy` or similar. Change the `kind` of the request span to `internal` if the frontend span is present.
1. Use a single span with `start_timestamp: request_start_time, kind: :server` and add a timed event for the service start time (i.e. "now").
1. Create both spans as children of the `extracted_context`, with the frontend span having `start_timestamp: request_start_time, end_timestamp: now` and the request span having `start_timestamp: now`. This allows the frontend span to represent queuing time only.
1. Create 2 spans for the frontend: a `kind: :server` and a `kind: :client` with the same start time and duration but a different resource. Specifying a different resource is not supported by the API presently.

Reply:
My vote would be for option 1 or 2. The original datadog instrumentation did have `span.kind` = `proxy`. `internal` would work for now and we can change it to `proxy` when and if it gets added.

As a side issue, I think we are going to have problems with nested server spans. Currently Sinatra will set its span kind as `server`. Using Rack with Sinatra will result in nesting of server spans. We will need to come up with a way to resolve this issue as a user could use Sinatra without Rack, and that would be what you want.

The same is true for client spans with Faraday + net http.

Reply:
I've updated, using Option 1.

Reply:
I'm happy with this. @fbogsany WDYT?